File: automation-test-playwright/pages/customer/CustomerServicePage.ts
import { type Locator, type Page } from "@playwright/test";
import { CustomerShellPage } from "../core/CustomerShellPage";

export class CustomerServicePage extends CustomerShellPage {
  protected readonly path = "/customer/service";
  readonly categories: Locator;
  private readonly cards: Locator;

  /**
   * Initializes this page object.
   */
  constructor(page: Page) {
    super(page);
    this.categories = this.page.locator(".category-section");
    this.cards = this.page.locator(".service-card");
  }

  async waitForLoaded(): Promise<void> {
    await this.page.waitForURL(/\/customer\/service/);
    await this.waitForTitleContainsLoose("dich vu", "suntower");
    await this.waitForLocatorCount(this.categories, 2);
  }

  async cardByTitle(title: string): Promise<Locator> {
    const expectedWords = this.serviceTitleWords(title);
    const count = await this.cards.count();

    for (let index = 0; index < count; index += 1) {
      const card = this.cards.nth(index);
      const normalizedText = this.normalizeLooseText(await card.innerText());
      if (expectedWords.every((word) => normalizedText.includes(word))) {
        return card;
      }
    }

    return this.firstVisible(this.cards.filter({ hasText: title }));
  }

  async waitForCardVisible(title: string): Promise<void> {
    await this.waitForVisible(await this.cardByTitle(title));
  }

  async waitForRequestDisabled(title: string): Promise<void> {
    await this.waitForDisabled((await this.cardByTitle(title)).locator(".btn-request"));
  }

  private serviceTitleWords(title: string): string[] {
    const normalizedTitle = this.normalizeLooseText(title);
    const aliases: Array<{ includes: string[]; words: string[] }> = [
      { includes: ["xe"], words: ["xe", "to"] },
      { includes: ["internet"], words: ["internet"] },
      { includes: ["gym"], words: ["gym"] },
      { includes: ["an ninh"], words: ["an", "ninh"] }
    ];
    return (
      aliases.find((alias) => alias.includes.every((word) => normalizedTitle.includes(word)))?.words ??
      normalizedTitle.split(" ")
    );
  }
}

File: automation-test-selenium/src/main/java/org/suntower/pages/customer/CustomerServicePage.java
package org.suntower.pages.customer;

import java.util.Arrays;
import java.util.List;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.suntower.core.BasePage;

public class CustomerServicePage extends BasePage {
  private final By categories = css(".category-section");
  private final By cards = css(".service-card");

  public CustomerServicePage(WebDriver driver) {
    super(driver);
  }

  public void waitForLoaded() {
    waitForUrlContains("/customer/service");
    waitForCondition(() -> driver.findElements(categories).size() >= 2, "Customer service categories did not load.");
  }

  public void waitForCardVisible(String title) {
    cardByTitle(title);
  }

  public void waitForRequestDisabled(String title) {
    WebElement card = cardByTitle(title);
    waitForCondition(
        () -> card.findElements(By.cssSelector(".btn-request")).stream().anyMatch(button -> !button.isEnabled() || button.getAttribute("disabled") != null),
        "Request button was not disabled for service: " + title);
  }

  private WebElement cardByTitle(String title) {
    List<String> words = serviceTitleWords(title);
    waitForCondition(
        () ->
            driver.findElements(cards).stream()
                .anyMatch(card -> card.isDisplayed() && words.stream().allMatch(word -> normalizeLooseText(card.getText()).contains(word))),
        "No service card contained title words: " + title);
    return driver.findElements(cards).stream()
        .filter(card -> card.isDisplayed() && words.stream().allMatch(word -> normalizeLooseText(card.getText()).contains(word)))
        .findFirst()
        .orElseThrow();
  }

  private List<String> serviceTitleWords(String title) {
    String normalized = normalizeLooseText(title);
    if (normalized.contains("xe")) return List.of("xe");
    if (normalized.contains("internet")) return List.of("internet");
    if (normalized.contains("gym")) return List.of("gym");
    if (normalized.contains("an ninh")) return List.of("an", "ninh");
    return Arrays.asList(normalized.split(" "));
  }
}
